src/Schema/Directives/BaseDirective.php

```diff
@@ -292,4 +301,101 @@ protected function loadArgValues(): void
             $this->directiveArgs[$node->name->value] = AST::valueFromASTUntyped($node->value);
         }
     }
+
+    /**
+     * Get an instance of a directive.
+     *
+     * @api
+     */
+    public static function getDirectiveInstance(string $directiveSource, ScalarTypeDefinitionNode|ScalarTypeExtensionNode|ObjectTypeDefinitionNode|ObjectTypeExtensionNode|InterfaceTypeDefinitionNode|InterfaceTypeExtensionNode|UnionTypeDefinitionNode|UnionTypeExtensionNode|EnumTypeDefinitionNode|EnumTypeExtensionNode|InputObjectTypeDefinitionNode|InputObjectTypeExtensionNode|FieldDefinitionNode|InputValueDefinitionNode|EnumValueDefinitionNode $node): Directive
+    {
+        $directiveNode = Parser::directive($directiveSource);
+        $node->directives[] = $directiveNode;
+
+        $directiveLocator = Container::getInstance()->make(DirectiveLocator::class);
+        $directiveInstance = $directiveLocator->create($directiveNode->name->value);
+        if ($directiveInstance instanceof BaseDirective) {
+            $directiveInstance->hydrate($directiveNode, $node);
+        }
+
+        return $directiveInstance;
+    }
+
+    /**
+     * Add a directive to a type definition dynamically.
+     *
+     * @api
+     */
+    public static function addDirectiveToTypeDefinition(
+        string $directiveSource,
+        DocumentAST &$documentAST,
+        TypeDefinitionNode &$typeDefinition,
+    ): void {
+        if (!($typeDefinition instanceof ScalarTypeDefinitionNode ||
+            $typeDefinition instanceof ScalarTypeExtensionNode
+            || $typeDefinition instanceof ObjectTypeDefinitionNode
+            || $typeDefinition instanceof ObjectTypeExtensionNode
+            || $typeDefinition instanceof InterfaceTypeDefinitionNode
+            || $typeDefinition instanceof InterfaceTypeExtensionNode
+            || $typeDefinition instanceof UnionTypeDefinitionNode
+            || $typeDefinition instanceof UnionTypeExtensionNode
+            || $typeDefinition instanceof EnumTypeDefinitionNode
+            || $typeDefinition instanceof EnumTypeExtensionNode
+            || $typeDefinition instanceof InputObjectTypeDefinitionNode
+            || $typeDefinition instanceof InputObjectTypeExtensionNode
+            || $typeDefinition instanceof FieldDefinitionNode
+            || $typeDefinition instanceof InputValueDefinitionNode
+            || $typeDefinition instanceof EnumValueDefinitionNode)) {
+            throw new DefinitionException("The dynamically added directive [{$directiveSource}] can only be added to nodes that support directives.");
+        }
+
+        $directiveInstance = self::getDirectiveInstance($directiveSource, $typeDefinition);
+
+        if (!($directiveInstance instanceof TypeManipulator)) {
+            throw new DefinitionException("The dynamically added directive [{$directiveSource}] must implement the TypeManipulator interface.");
```

Comment:
I don't see why this would need to be the case to add a directive to a type definition.

I am suspicious if this method is worth adding, perhaps calling `manipulate*Definition` can just be done ad-hoc.

Reply:
Yep, sorry, you are right I just dropped it.

Reply:
Some with the other two functions like this one.

Reply:
Sorry, could you please elaborate on this?